source/internal/_clone.js

```diff
@@ -12,29 +11,104 @@ import type from '../type';
  * @param {Boolean} deep Whether or not to perform deep cloning.
  * @return {*} The copied value.
  */
-export default function _clone(value, refFrom, refTo, deep) {
+export default function _clone(value, deep, map) {
+
+  map || (map = new _ObjectMap());
+
+  // this avoids the slower switch with a quick if decision removing some milliseconds in each run.
+  if (_isPrimitive(value)) {
+    return value;
+  }
+
   var copy = function copy(copiedValue) {
-    var len = refFrom.length;
-    var idx = 0;
-    while (idx < len) {
-      if (value === refFrom[idx]) {
-        return refTo[idx];
-      }
-      idx += 1;
+    // Check for circular and same references on the object graph and return its corresponding clone.
+    var cachedCopy = map.get(value);
+
+    if (cachedCopy) {
+      return cachedCopy;
     }
-    refFrom[idx] = value;
-    refTo[idx] = copiedValue;
+
+    map.set(value, copiedValue);
+
     for (var key in value) {
-      copiedValue[key] = deep ?
-        _clone(value[key], refFrom, refTo, true) : value[key];
+      copiedValue[key] = deep ? _clone(value[key], true, map) : value[key];
     }
     return copiedValue;
   };
+
   switch (type(value)) {
-    case 'Object':  return copy({});
-    case 'Array':   return copy([]);
-    case 'Date':    return new Date(value.valueOf());
-    case 'RegExp':  return _cloneRegExp(value);
-    default:        return value;
+    case 'Object':
+      return copy({});
+    case 'Array':
+      return copy([]);
+    case 'Date':
+      return new Date(value.valueOf());
+    case 'RegExp':
+      return _cloneRegExp(value);
+    default:
+      return value;
   }
 }
+
+function _isPrimitive(param) {
+  var type = typeof param;
+  return param == null || (type != 'object' && type != 'function');
+}
+
+function _ObjectMap() {
+  this.map = {};
+  this.length = 0;
+}
+
+_ObjectMap.prototype.set = function(key, value) {
+  const hashedKey = this.hash(key);
+
+  let bucket = this.map[hashedKey];
+  if (!bucket) {
+    this.map[hashedKey] = bucket = [];
+  }
+
+  bucket.push([key, value]);
+  this.length += 1;
+};
+
+_ObjectMap.prototype.hash = function(key) {
+  let hashedKey = [];
+  for (var value in key) {
+    hashedKey.push(key[value]);
+  }
+  return hashedKey.join();
+};
+
+_ObjectMap.prototype.get = function(key) {
+
+  /**
+   * depending on the number of objects to be cloned is faster to just iterate over the items in the map just because the hash function is so costly,
+   * on my tests this number is 180, anything above that using the hash function is faster.
+   */
+  if (this.length <= 180) {
+
+    for (const p in this.map) {
+      const bucket = this.map[p];
+
+      for (let i = 0; i < bucket.length; i += 1) {
+        const element = bucket[i];
+        if (element[0] === key) {return element[1];}
```

Comment:
I think this would have to be `if (equals (element [0], key)) ...`

Ramda is all about value equality, not reference equality.

Reply:
While I understand that Ramda is about value equality I think that the clone function should preserve the structure of the object, let's say:
```js
const d = {foo:1, bar:"2"};
const a = {foo: d, c:{ bar: d}};
a.foo === a.c.bar // == true;

//the clone method should preserve the structure.

const clone = R.clone(a);
clone.foo === clone.c.bar // == should also be true;
```
That's the way the R.clone() is behaving and lodash cloneDeep() also behave like this and I also believe it should behave like this.

Reply:
I suppose you're right that this is how it's working now.  I never worked closely with this code and was thinking that we had the stronger guarantee of preserving the value equality structure (as if the above with `x === y` replaced with `equals (x, y)`).  Part of me would still like that.  But that would further degrade performance, not enhance it!

Reply:
But `(x === y); //=== true` then `equals(x,y) //must also be true`. So by preserving reference equality we are also preserving value equality.

Reply:
I was suggesting that we might go further than that, but on further reflection I realized that it's not only unnecessary but actively harmful to the goal here.